File: backend/schemas/machine_auth.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class InviteCodeCreate(BaseModel):
    # Scenario A (new photographer): set photographer_name + access_level
    photographer_name: str | None = None
    access_level: str = "guest"
    ttl_minutes: int = 60
    # Scenario B (existing photographer): set target_photographer_id
    target_photographer_id: uuid.UUID | None = None


class InviteCodeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    code: str
    access_level: str | None
    target_photographer_id: uuid.UUID | None
    photographer_name: str | None
    expires_at: datetime
    used_at: datetime | None
    created_at: datetime


class EnrollRequest(BaseModel):
    code: str
    device_name: str = ""


class EnrollResponse(BaseModel):
    machine_id: uuid.UUID
    api_token: str
    photographer_id: uuid.UUID
    photographer_name: str


class AddMachineCodeResponse(BaseModel):
    code: str
    expires_at: datetime


class MachineOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    machine_id: uuid.UUID
    machine_name: str
    photographer_id: uuid.UUID | None
    last_seen_at: datetime | None
    created_at: datetime


class PhotographerWithMachinesOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    access_level: str
    machines: list[MachineOut] = []

File: frontend/src/api/machineAuth.ts
import { apiFetch } from './client'

export interface InviteCode {
  id: string
  code: string
  access_level: string | null
  target_photographer_id: string | null
  photographer_name: string | null
  expires_at: string
  used_at: string | null
  created_at: string
}

export interface MachineOut {
  machine_id: string
  machine_name: string
  photographer_id: string | null
  last_seen_at: string | null
  created_at: string
}

export interface PhotographerWithMachines {
  id: string
  name: string
  access_level: string
  machines: MachineOut[]
}

export function createInviteCode(body: {
  photographer_name?: string
  access_level?: string
  ttl_minutes?: number
  target_photographer_id?: string
}): Promise<InviteCode> {
  return apiFetch('/admin/invite-codes', { method: 'POST', body: JSON.stringify(body) })
}

export function listInviteCodes(): Promise<InviteCode[]> {
  return apiFetch('/admin/invite-codes')
}

export function deleteInviteCode(id: string): Promise<void> {
  return apiFetch(`/admin/invite-codes/${id}`, { method: 'DELETE' })
}

export function listMachinesAdmin(): Promise<MachineOut[]> {
  return apiFetch('/admin/machines')
}

export function revokeMachineToken(machineId: string): Promise<void> {
  return apiFetch(`/admin/machines/${machineId}/token`, { method: 'DELETE' })
}

export function listPhotographersWithMachines(): Promise<PhotographerWithMachines[]> {
  return apiFetch('/admin/photographers')
}

export function setPhotographerAccessLevel(photographerId: string, accessLevel: string): Promise<void> {
  return apiFetch(`/admin/photographers/${photographerId}/access-level`, {
    method: 'PATCH',
    body: JSON.stringify({ access_level: accessLevel }),
  })
}
